keras_nlp/utils/text_generation.py

```diff
@@ -703,22 +728,20 @@ def token_probability_fn(inputs):
             f"`p` should be in the range (0, 1). Received: `p={p}`."
         )
 
-    prompt = validate_prompt(prompt)
+    prompt = _validate_prompt(prompt)
+
     input_is_1d = prompt.shape.rank == 1
     if input_is_1d:
         prompt = prompt[tf.newaxis, :]
-    validate_token_probability_fn(token_probability_fn, prompt)
 
-    shape = tf.shape(prompt)
-    batch_size = shape[0]
-    length = shape[1]
+    _validate_token_probability_fn(token_probability_fn, prompt)
 
-    # Pad the prompt with `pad_token_id` to `max_length`.
-    padding = tf.fill((batch_size, max_length - length), pad_token_id)
-    prompt = tf.concat((prompt, padding), axis=1)
+    prompt, mask, batch_size, length = _pad_prompts(
```

Comment:
piggybacking on above comment re using ragged bounding shape, maybe we should just split this in two?

```
batch_size, length = _get_prompt_bounding_shape(prompt)
prompt, mask = _pad_prompt(prompt, max_length, pad_token_id)
```

For now, you will still need to read the shape inside `_pad_prompt` for the dense branch, but that's a temporary workaround anyway.

Reply:
For now, I've done this:

```    
    batch_size, length = _get_prompt_length(prompt)
    prompt, mask = _pad_prompt(
        prompt, batch_size, length, max_length, pad_token_id
    )
```

Reply:
Still think this could be simplified a bit, these are complex chains of ops and readability is important.

I'll push a commit with a suggestion, take a look!